src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -2333,192 +1456,100 @@ class HandshakeForDeflation : public HandshakeClosure {
+    // Unlink deflated ObjectMonitors from the in-use list.
+    ResourceMark rm;
+    GrowableArray<ObjectMonitor*> delete_list((int)deflated_count);
+    size_t unlinked_count = _in_use_list.unlink_deflated(self, ls, &timer,
+                                                         &delete_list);
+    if (self->is_Java_thread()) {
+      if (ls != NULL) {
+        timer.stop();
+        ls->print_cr("before handshaking: unlinked_count=" SIZE_FORMAT
+                     ", in_use_list stats: ceiling=" SIZE_FORMAT ", count="
+                     SIZE_FORMAT ", max=" SIZE_FORMAT,
+                     unlinked_count, in_use_list_ceiling(),
+                     _in_use_list.count(), _in_use_list.max());
+      }
 
-      OM_PERFDATA_OP(Deflations, inc(local_deflated_count));
-    }
+      // A JavaThread needs to handshake in order to safely free the
+      // ObjectMonitors that were deflated in this cycle.
+      HandshakeForDeflation hfd_hc;
+      Handshake::execute(&hfd_hc);
 
-    if (saved_mid_in_use_p != NULL) {
-      // deflate_monitor_list() detected a safepoint starting.
-      timer.stop();
-      {
-        if (is_global) {
-          log_debug(monitorinflation)("pausing deflation of global idle monitors for a safepoint.");
-        } else {
-          log_debug(monitorinflation)("jt=" INTPTR_FORMAT ": pausing deflation of per-thread idle monitors for a safepoint.", p2i(target));
-        }
-        assert(self->is_Java_thread() &&
-               SafepointMechanism::should_process(self->as_Java_thread()),
-               "sanity check");
-        ThreadBlockInVM blocker(self->as_Java_thread());
-      }
-      // Prepare for another loop after the safepoint.
-      free_head_p = NULL;
-      free_tail_p = NULL;
-      if (log_is_enabled(Info, monitorinflation)) {
+      if (ls != NULL) {
+        ls->print_cr("after handshaking: in_use_list stats: ceiling="
+                     SIZE_FORMAT ", count=" SIZE_FORMAT ", max=" SIZE_FORMAT,
+                     in_use_list_ceiling(), _in_use_list.count(), _in_use_list.max());
         timer.start();
       }
     }
-  } while (saved_mid_in_use_p != NULL);
-  timer.stop();
 
-  LogStreamHandle(Debug, monitorinflation) lsh_debug;
-  LogStreamHandle(Info, monitorinflation) lsh_info;
-  LogStream* ls = NULL;
-  if (log_is_enabled(Debug, monitorinflation)) {
-    ls = &lsh_debug;
-  } else if (deflated_count != 0 && log_is_enabled(Info, monitorinflation)) {
-    ls = &lsh_info;
+    // After the handshake, safely free the ObjectMonitors that were
+    // deflated in this cycle.
+    size_t deleted_count = 0;
+    for (ObjectMonitor* monitor: delete_list) {
+      delete monitor;
+      deleted_count++;
+
+      if (self->is_Java_thread()) {
+        // A JavaThread must check for a safepoint/handshake and honor it.
+        chk_for_block_req(self->as_Java_thread(), "deletion", "deleted_count",
+                          deleted_count, ls, &timer);
```

Comment:
If you release oopStorage when deflating you can do this entire loop while blocked instead, which will be faster.

Reply:
@fisk said we should not release the oopStorage during a safepoint
because that's not safe or will not be safe. I can't remember which.

Reply:
Yes that's why I said you can release it during deflation instead.

Reply:
What does this "you can do this entire loop while blocked instead" mean?

Releasing during deflation kind of messes with the life-cycle I was trying
to enforce since deletion is the nature end-of-life for these... But to think
about that I need to know what you mean by "you can do this entire loop
while blocked instead"...

Reply:
If you only need to free CHeap memory, you can do:
```
   size_t deleted_count = 0;
   ThreadBlockInVM tbivm(self);
    for (ObjectMonitor* monitor: delete_list) {
      delete monitor;
      deleted_count++;
    }
  }
```

Reply:
Ahhh... but that only works if we release the oopStorage when
we deflate. Okay. I grok it now, but don't want to do that in this
changeset. I would want a complete stress test cycle for that
kind of a change.

Reply:
Filed this new RFE: